manual/src/refman/extensions/tail_mod_cons.etex

```diff
@@ -0,0 +1,548 @@
+
+type exp =
+  | Var of var
+  | Let of binding * exp
+and binding = var * exp
+\end{verbatim}
+
+Consider a "map" function on variables. The direct definition has two
+recursive calls inside arguments of the "Let" constructor, so it gets
+rejected as ambiguous.
+\begin{verbatim}
+let[@tail_mod_cons] rec map_vars f exp =
+  match exp with
+  | Var v -> Var (f v)
+  | Let ((v, def), body) ->
+    Let ((f v, map_vars f def), map_vars f body)
+
+Error: [@tail_mod_cons]: this constructor application may be TMC-transformed
+       in several different ways. Please disambiguate by adding an explicit
+       [@tailcall] attribute to the call that should be made tail-recursive,
+       or a [@tailcall false] attribute on calls that should not be
+       transformed.
+Line 5, characters 15-29:   This call could be annotated.
+Line 5, characters 32-47:   This call could be annotated.
+\end{verbatim}
+
+To disambiguate, the user should add a "[\@tailcall]" attribute to the
+recursive call that they want to see in tail position:
+\begin{verbatim}
+let[@tail_mod_cons] rec map_vars f exp =
+  match exp with
+  | Var v -> Var (f v)
+  | Let ((v, def), body) ->
+    Let ((f v, map_vars f def), (map_vars[@tailcall]) f body)
+\end{verbatim}
+Beware that the resulting function is \emph{not} tail-recursive, the
+recursive call on "def" will consume stack space. However, expression
+trees tend to unbalanced on the right (lots of "Let" in sequence,
+rather than nested inside each other), so putting the call on "body"
+in tail position is an interesting improvement over the naive
+definition: it gives bounded stack space consumption if we assume
+a bound on the nesting depth of "Let" constructs.
+
+One would also get an error when using conflicting annotations, asking
+for two of the constructor arguments to be put in tail position:
+\begin{verbatim}[error]
+let[@tail_mod_cons] rec map_vars f exp =
+  match exp with
+  | Var v -> Var (f v)
+  | Let ((v, def), body) ->
+    Let ((f v, (map_vars[@tailcall]) f def), (map_vars[@tailcall]) f body)
+
+Error: [@tail_mod_cons]: this constructor application may be TMC-transformed
+       in several different ways. Only one of the arguments may become a TMC
+       call, but several arguments contain calls that are explicitly marked
+       as tail-recursive. Please fix the conflict by reviewing and fixing the
+       conflicting annotations.
+Line 5, characters 15-42:   This call is explicitly annotated.
+Line 5, characters 45-73:   This call is explicitly annotated.
+\end{verbatim}
+
+\subsection{ss:out-of-tmc}{Danger: getting out of tail-mod-cons}
+
+Due to the nature of the tail-mod-cons transformation
+(see Section~\ref{ss:details} for a presentation of transformation):
+\begin{itemize}
+\item Calls from a tail-mod-cons function to another tail-mod-cons
+  function declared in the same recursive-binding group are ``tail
+  calls'', whether they occur in tail position or in tail-modulo-cons
```

Comment:
Is there something wrong with the "whether .. or" construct here? This is what I wanted to use and it seems clearer.

Reply:
It might be misread as "whether they occur in (tail position or tail-modulo-cons position) or not".
Most often (always?) in "whether A or B" the A and B exhaust all possibilities (often B is the negation or A, and often it might be written "or not" or even omitted). This is not the case here, as a call might be neither tail position nor tail-modulo-cons position.

Reply:
One could further emphasise this by having "_even_ if they occur"

Reply:
Umm... "*only* if"?
```suggestion
  calls'', only if they occur in tail position or in tail-modulo-cons
```

Reply:
Yes, that’s what I meant 😳

Reply:
I rephrased as follows:

> \item Calls from a tail-mod-cons function to another tail-mod-cons
  function declared in the same recursive-binding group are
  transformed into tail calls, as soon as they occur in tail position
  or tail-modulo-cons position in the source function.
\item Calls from a function \emph{not} marked tail-mod-cons to
  a tail-mod-cons function or, conversely, from a tail-mod-cons
  function to a non-tail-mod-cons function are transformed into
  \emph{non}-tail calls, even if they syntactically appear in tail
  position in the source program.